src/middlewared/middlewared/api/base/server/ws_handler/rpc.py

```diff
@@ -0,0 +1,301 @@
+            "frames": frames,
+            "formatted": "".join(traceback.format_exception(*exc_info)),
+            "repr": repr(value),
+        }
+
+    def send_truenas_validation_error(self, id_: Any, exc_info, errors: list):
+        self.send_error(id_, JSONRPCError.INVALID_PARAMS.value, "Invalid params",
+                        self.format_truenas_validation_error(exc_info[1], exc_info, errors))
+
+    def format_truenas_validation_error(self, exception, exc_info=None, errors: list | None = None):
+        return self.format_truenas_error(errno.EINVAL, str(exception), exc_info, errors)
+
+    def register_callback(self, event: RpcWebSocketAppEvent, callback: Callable):
+        self.callbacks[event.value].append(callback)
+
+    def run_callback(self, event, *args, **kwargs):
+        for callback in self.callbacks[event.value]:
+            try:
+                callback(self, *args, **kwargs)
+            except Exception:
+                logger.error(f"Failed to run {event} callback", exc_info=True)
+
+    async def subscribe(self, ident: str, name: str):
+        shortname, arg = self.middleware.event_source_manager.short_name_arg(name)
+        if shortname in self.middleware.event_source_manager.event_sources:
+            await self.middleware.event_source_manager.subscribe_app(self, self.__esm_ident(ident), shortname, arg)
+        else:
+            self.subscriptions[ident] = name
+
+    async def unsubscribe(self, ident: str):
+        if ident in self.subscriptions:
+            self.subscriptions.pop(ident)
+        elif self.__esm_ident(ident) in self.middleware.event_source_manager.idents:
+            await self.middleware.event_source_manager.unsubscribe(self.__esm_ident(ident))
+
+    def __esm_ident(self, ident):
+        return self.session_id + ident
+
+    def send_event(self, name: str, event_type: str, **kwargs):
+        if (
+            not any(i in [name, "*"] for i in self.subscriptions.values()) and
+            (
+                self.middleware.event_source_manager.short_name_arg(name)[0] not in
+                self.middleware.event_source_manager.event_sources
+            )
+        ):
+            return
+
+        event = {
+            "msg": event_type.lower(),
+            "collection": name,
+        }
+        kwargs = kwargs.copy()
+        if "id" in kwargs:
+            event["id"] = kwargs.pop("id")
+        if event_type in ("ADDED", "CHANGED"):
+            if "fields" in kwargs:
+                event["fields"] = kwargs.pop("fields")
+        if kwargs:
+            event["extra"] = kwargs
+
+        self.send_notification("collection_update", event)
+
+    def notify_unsubscribed(self, collection: str, error: Exception | None):
+        params = {"collection": collection, "error": None}
+        if error:
+            if isinstance(error, ValidationErrors):
+                params["error"] = self.format_truenas_validation_error(error, extra=list(error))
```

Comment:
`format_truenas_validation_error` doesn't have a kwarg `extra`, maybe you meant `errors`?